File: flutter/lib/models/playlist.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';
import 'music_track.dart';

part 'playlist.g.dart';

enum PlaylistType {
  daily,
  weekly,
  planetBased,
  moodBased,
  custom,
  transit,
}

@JsonSerializable()
class Playlist extends Equatable {
  final String id;
  final String title;
  final String description;
  final PlaylistType type;
  final String? userId;
  final List<String> trackIds;
  final String? coverImageUrl;
  final DateTime createdAt;
  final DateTime? expiresAt; // For daily/weekly playlists
  final Map<String, dynamic>? astrologicalContext; // Planet positions, transits, etc.
  final bool isPersonalized;
  final List<String> tags;

  const Playlist({
    required this.id,
    required this.title,
    required this.description,
    required this.type,
    this.userId,
    required this.trackIds,
    this.coverImageUrl,
    required this.createdAt,
    this.expiresAt,
    this.astrologicalContext,
    this.isPersonalized = false,
    this.tags = const [],
  });

  factory Playlist.fromJson(Map<String, dynamic> json) =>
      _$PlaylistFromJson(json);

  Map<String, dynamic> toJson() => _$PlaylistToJson(this);

  bool get isExpired {
    if (expiresAt == null) return false;
    return DateTime.now().isAfter(expiresAt!);
  }

  int get trackCount => trackIds.length;

  Playlist copyWith({
    String? id,
    String? title,
    String? description,
    PlaylistType? type,
    String? userId,
    List<String>? trackIds,
    String? coverImageUrl,
    DateTime? createdAt,
    DateTime? expiresAt,
    Map<String, dynamic>? astrologicalContext,
    bool? isPersonalized,
    List<String>? tags,
  }) {
    return Playlist(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      type: type ?? this.type,
      userId: userId ?? this.userId,
      trackIds: trackIds ?? this.trackIds,
      coverImageUrl: coverImageUrl ?? this.coverImageUrl,
      createdAt: createdAt ?? this.createdAt,
      expiresAt: expiresAt ?? this.expiresAt,
      astrologicalContext: astrologicalContext ?? this.astrologicalContext,
      isPersonalized: isPersonalized ?? this.isPersonalized,
      tags: tags ?? this.tags,
    );
  }

  @override
  List<Object?> get props => [
        id,
        title,
        description,
        type,
        userId,
        trackIds,
        coverImageUrl,
        createdAt,
        expiresAt,
        astrologicalContext,
        isPersonalized,
        tags,
      ];
}

File: backend/app/models/playlist.py
"""
Playlist database model
"""
from sqlalchemy import Column, String, Text, DateTime, ForeignKey, Boolean, JSON, Enum as SQLEnum
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid
import enum
from app.db.base import Base


class PlaylistType(str, enum.Enum):
    """Playlist type enumeration"""
    DAILY = "daily"
    WEEKLY = "weekly"
    PLANET_BASED = "planet_based"
    MOOD_BASED = "mood_based"
    CUSTOM = "custom"
    TRANSIT = "transit"


class Playlist(Base):
    """Playlist model"""

    __tablename__ = "playlists"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    title = Column(String, nullable=False)
    description = Column(Text)

    # Associations
    user_id = Column(String, ForeignKey("users.id"), nullable=True)  # User playlists
    playlist_type = Column(SQLEnum(PlaylistType), nullable=False)

    # Content
    track_ids = Column(JSON, default=list)  # List of track IDs
    cover_image_url = Column(String)

    # Metadata
    is_personalized = Column(Boolean, default=False)
    tags = Column(String)  # Comma-separated
    astrological_context = Column(JSON)  # Planet positions, transits, etc.

    # Expiration (for daily/weekly playlists)
    expires_at = Column(DateTime, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    user = relationship("User", back_populates="playlists")
